src/pymor/algorithms/cholesky_qr.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python3
+
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.algorithms.eigs import eigs
+from pymor.core.logger import getLogger
+
+logger = getLogger('pymor.algorithms.cholesky_qr.cholesky_qr')
+
+
+def _chol_qr(A, product=None, gramian=None, check_finite=True):
+    X = A.gramian(product=product) if gramian is None else gramian
+
+    while True:
+        try:
+            R = spla.cholesky(X, overwrite_a=True, check_finite=check_finite)
+            break
+        except spla.LinAlgError:
+            logger.warning('Cholesky factorization broke down! Matrix is ill-conditioned.')
+            m, n = A.dim, len(A)
+            eps = np.finfo(X.dtype).eps
+            s = m*n+n*(n+1) if product is None else 2*m*np.sqrt(m*n)+n*(n+1)*np.sqrt(np.abs(eigs(product, k=1)[0][0]))
+            s *= 11*eps*spla.norm(X, ord=2, check_finite=check_finite)
```

Comment:
At the moment, I the shift is incorrect if `product` is not `None`.

I am using the fact that $\lVert A\rVert_2^2=\lVert A^TA\rVert_2$ so that I can use `scipy.norm` on the gramian.

However, with a product I need a good way to compute $\lVert A\rVert_2^2\lVert B\rVert_2$. For now, I am computing $\lVert A^TBA\rVert_2\lVert B\rVert_2$ but thats not correct...

Reply:
I don't see an easy way around computing $\lVertAˆTA\rVert_2^2$ so I just do that. It is not too expensive after all...